### ISSUE TITLE
ORB exception 

### ISSUE THREAD

**Issue by kahrabji08** [arrayfire/arrayfire#1531]
Am trying to use ORB function . My code is huge, here is what i think is relevant 

``` cpp
    if(_image.empty())
        return;

    Mat image = _image.getMat(), mask = _mask.getMat();
    assert(image.type() == CV_8UC1 );   
    array imArray;
    mat_to_array_(image,imArray); \\convert Opencv Mat to Arrayfire Array    

    if (imArray.isempty())
        return; 

    features orb_features;
    array orb_descriptors;
    orb(orb_features, orb_descriptors, imArray, fastTh, nfeatures, scaleFactor, nlevels, true);
```

The program runs fine mostly but sometimes I get the following error:

```
terminate called after throwing an instance of 'af::exception'
  what():  ArrayFire Exception (Unknown error:999):

In function void af::orb(af::features&, af::array&, const af::array&, float, unsigned int, float, unsigned int, bool)
In file src/api/cpp/orb.cpp:25
```

An sometimes I get this

```
terminate called after throwing an instance of 'af::exception'
  what():  ArrayFire Exception (Device out of memory:101):
In function nativeAlloc
In file src/backend/cuda/memory.cpp:97
CUDA Error (2): out of memory


In function void af::orb(af::features&, af::array&, const af::array&, float, unsigned int, float, unsigned int, bool)
In file src/api/cpp/orb.cpp:25

```

**Comment by 9prady9** [comment]
@kahrabji08 I still an issue, can you please retry your code against our latest release - ArrayFire v3.6.

**Comment by mark-poscablo** [comment]
@kahrabji08 Also do you remember which (if not all) backends does this happen?

**Comment by MohameHassan** [comment]
Hi,
I am sorry but I don't remember any details about this issue.

**Comment by 9prady9** [comment]
Could be related to https://github.com/arrayfire/arrayfire/pull/2177 but we will confirm that and update here. @mark-poscablo  Can you please run your test code against master after 2177 is merged. Thank you.

**Comment by mark-poscablo** [comment]
@9prady9 building it now on my OS X machine and will let you guys know how it fares on my sample code.

Btw I also looked at the backtraces and they suggest that the errors happen somewhere in the windowing system and graphics. The errors happen to be about double freeing pointers and bad accesses. So yeah I agree with you that this might be related to memory leaks and such

**Comment by mark-poscablo** [comment]
Unfortunately segfault's still there on my OS X, OpenCL

**Comment by 9prady9** [comment]
@mark-poscablo Which device are you using on OSX OpenCL ? Can you test this on Linux.

**Comment by mark-poscablo** [comment]
@9prady9 On OS X I'm just using a simple Intel Iris Graphics 6100 GPU. On Linux (Ubuntu), I do have an Nvidia Geforce 940M (not GTX) and there all backends produce the same error: double freeing a pointer, but only when I try to display the output result in a window and close it.

**Comment by 9prady9** [comment]
thats fine, I have GT650M, let me try on that and see where the double free is stemming from. Thanks!

**Comment by mark-poscablo** [comment]
no prob! let me know what you find out - I'm going to try and poke around it again now that I'm back from training. btw feel free to look at the mini-report on google drive that I shared with you about this error - it contains my sample code and the backtrace info when I tried running gdb on my sample code. thanks too!

**Comment by mark-poscablo** [comment]
So we've been working on this and it looks like the memory leaks happen on the glbinding side. In my case, using my Ubuntu machine, I had glbinding 2.1.3 install through apt-get, and that's when I kept getting errors about double-freeing a pointer after I close the output window. I also used address sanitizer (by adding `-fsanitize=address -fno-omit-frame-pointer`) and it did not show memory leaks caused by the arrayfire library itself, as vetted too by @9prady9 . In the end I tried to build glbinding 2.1.4 from source and linked arrayfire to it before rebuilding arrayfire. Then I tried to run my code sample again on all backends and they worked fine - there were no more errors about double-freeing a pointer.

In summary, the issue looks like it's fixed by the newer version of glbinding (2.1.4), so if anyone encounters this issue again, I'd recommend using glbinding >= 2.1.4 to resolve the issue.
